File: app/src/main/java/nl/jamiecraane/demo/recyclerviewviewpager/ScrollListener.java
package nl.jamiecraane.demo.recyclerviewviewpager;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import rx.Observable;
import rx.subjects.BehaviorSubject;

import java.util.concurrent.TimeUnit;

/**
 * Created by jamiecraane on 31/07/2017.
 */

public class ScrollListener extends RecyclerView.OnScrollListener {
    private final LinearLayoutManager layoutManager;
    private final BehaviorSubject<Integer> positionSubject = BehaviorSubject.create();

    public ScrollListener(final LinearLayoutManager layoutManager) {
        this.layoutManager = layoutManager;
    }

    @Override
    public void onScrolled(final RecyclerView recyclerView, final int dx, final int dy) {
        super.onScrolled(recyclerView, dx, dy);
        positionSubject.onNext(layoutManager.findFirstCompletelyVisibleItemPosition());
    }

    public Observable<Integer> getPositionObservable() {
        return positionSubject
                .distinctUntilChanged()
                .filter(position -> position >= 0)
                .debounce(350, TimeUnit.MILLISECONDS);
    }
}

File: app/src/main/java/nl/jamiecraane/demo/recyclerviewviewpager/ScrollListener.kt
package nl.jamiecraane.demo.recyclerviewviewpager

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import rx.Observable
import rx.subjects.BehaviorSubject
import java.util.concurrent.TimeUnit

/**
 * Created by jamiecraane on 02/02/2018.
 */
class ScrollListener(private val layoutManager: LinearLayoutManager) : RecyclerView.OnScrollListener() {
    private val positionSubject = BehaviorSubject.create<Int>()

    val positionObservable: Observable<Int>
        get() = positionSubject
                .distinctUntilChanged()
                .filter { position -> position >= 0 }
                .debounce(350, TimeUnit.MILLISECONDS)

    override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)
        positionSubject.onNext(layoutManager.findFirstCompletelyVisibleItemPosition())
    }
}